src/EventStore.Client.Streams/EventStoreClient.Read.cs

```diff
@@ -65,23 +71,114 @@ public partial class EventStoreClient {
 				cancellationToken);
 		}
 
-		private ReadStreamResult ReadStreamAsync(Direction direction,
-			string streamName, StreamPosition revision, long maxCount,
-			EventStoreClientOperationOptions operationOptions, bool resolveLinkTos = false,
-			UserCredentials? userCredentials = null, CancellationToken cancellationToken = default) {
-			return new ReadStreamResult(SelectCallInvoker, new ReadReq {
-					Options = new ReadReq.Types.Options {
-						ReadDirection = direction switch {
-							Direction.Backwards => ReadReq.Types.Options.Types.ReadDirection.Backwards,
-							Direction.Forwards => ReadReq.Types.Options.Types.ReadDirection.Forwards,
-							_ => throw InvalidOption(direction)
-						},
-						ResolveLinks = resolveLinkTos,
-						Stream = ReadReq.Types.Options.Types.StreamOptions.FromStreamNameAndRevision(streamName,
-							revision),
-						Count = (ulong)maxCount
+		/// <summary>
+		/// A class that represents the result of a read operation on a stream.
+		/// </summary>
+		public class ReadAllStreamResult : IAsyncEnumerable<ResolvedEvent> {
+			private readonly Channel<StreamMessage> _channel;
+
+			/// <summary>
+			/// 
+			/// </summary>
+			public Position? LastPosition { get; private set; }
+
+			/// <summary>
+			/// An <see cref="IAsyncEnumerable{StreamMessage}"/>.
+			/// </summary>
+			public IAsyncEnumerable<StreamMessage> Messages {
+				get {
+					return GetMessages();
+
+					async IAsyncEnumerable<StreamMessage> GetMessages() {
+						await foreach (var message in _channel.Reader.ReadAllAsync()) {
+							if (message is StreamMessage.LastAllStreamPosition(var position)) {
+								LastPosition = position;
+							}
+
+							yield return message;
+						}
+					}
+				}
+			}
+
+			internal ReadAllStreamResult(Func<CancellationToken, Task<CallInvoker>> selectCallInvoker, ReadReq request,
+				EventStoreClientSettings settings, EventStoreClientOperationOptions operationOptions,
+				UserCredentials? userCredentials, CancellationToken cancellationToken) {
+				var callOptions = EventStoreCallOptions.Create(settings, operationOptions, userCredentials,
+					cancellationToken);
+
+				_channel = Channel.CreateBounded<StreamMessage>(new BoundedChannelOptions(64) {
```

Comment:
Did @jageall suggest to reduce this to 1 and enable AllowSynchronousContinuations (probs same for the stream version below)

Reply:
.. i may have misrembered/been plain confused